include/fw_info.h

```diff
@@ -87,81 +69,199 @@ struct __packed fw_info {
-	OFFSET_CHECK_EXT_API(type, ext_api_flags,\
-				(CONFIG_FW_INFO_MAGIC_LEN + 4)); \
-	OFFSET_CHECK_EXT_API(type, ext_api_version,\
-				(CONFIG_FW_INFO_MAGIC_LEN + 8)); \
-	OFFSET_CHECK_EXT_API(type, ext_api_len,\
-				(CONFIG_FW_INFO_MAGIC_LEN + 12)); \
+	Z_GENERIC_SECTION(.ext_apis) \
+	const u8_t _CONCAT(name, _ext_api_counter) = 0xFF; \
 	BUILD_ASSERT_MSG((sizeof(type) % 4) == 0, \
-			"ext_api " #type " is not word-aligned"); \
-	extern const type name; \
-	Z_GENERIC_SECTION(.ext_apis) __attribute__((used)) \
-	const type * const _CONCAT(name, _ptr) = &name; \
-	__attribute__((used)) \
+			"EXT_API " #type " is not word-aligned"); \
+	Z_GENERIC_SECTION(.firmware_info.1) __attribute__((used)) \
 	const type name
 
 
-
+/* Macro for initializing struct fw_info_ext_api instances. */
 #define FW_INFO_EXT_API_INIT(id, flags, version, total_size) \
 	{ \
-		.magic = {EXT_API_MAGIC}, \
+		.magic = {EXT_API_MAGIC},\
+		.ext_api_len = total_size, \
 		.ext_api_id = id, \
 		.ext_api_flags = flags, \
 		.ext_api_version = version, \
-		.ext_api_len = total_size, \
 	}
 
-/* Shorthand for declaring function that will be exposed through an ext_api.
+
+/* Check and provide a pointer to a fw_info_ext_api structure.
+ *
+ * @return pointer if valid, NULL if not.
+ */
+static inline const struct fw_info_ext_api *fw_info_ext_api_check(
+							u32_t ext_api_addr)
+{
+	const struct fw_info_ext_api *ext_api;
+	const u32_t ext_api_magic[] = {EXT_API_MAGIC};
+
+	ext_api = (const struct fw_info_ext_api *)(ext_api_addr);
+	if (memcmp(ext_api->magic, ext_api_magic, CONFIG_FW_INFO_MAGIC_LEN)
+		== 0) {
+		return ext_api;
+	}
+	return NULL;
+}
+
+
+/* Shorthand for declaring function that will be exposed through an EXT_API.
  * This will define a function pointer type as well as declare the function.
  */
 #define EXT_API_FUNCTION(retval, name, ...) \
-	typedef retval (*name ## _t) (__VA_ARGS__); \
-	retval name (__VA_ARGS__)
+	typedef retval(*name ## _t)(__VA_ARGS__); \
+	retval name(__VA_ARGS__)
+
+
+/**
+ * A struct that is used to request an EXT_API.
+ *
+ * @details It contains a pointer to a non-initialized pointer that a previous
+ *          boot stage will populate with a pointer to a compliant EXT_API. The
+ *          EXT_API could be located in the boot stage itself, or in a third
+ *          image. An EXT_API fulfills a request if the ID matches, all flags in
+ *          the request are set in the EXT_API, and the version falls between
+ *          the minimum and maximum (inclusive).
+ *
+ *          The request is placed in the list of requests in the fw_info struct.
+ *
+ * @note If `required` is true, the image making the request will not function
+ *       unless it has access to the EXT_API.
+ *
+ */
+struct __packed fw_info_ext_api_request {
+	/* The minimum EXT_API version. */
+	struct fw_info_ext_api min_ext_api;
+
+	/* The maximum version accepted. */
+	u32_t ext_api_max_version;
```

Comment:
Could you explain why the min version is a struct while the max version is an int?

Reply:
`min_ext_api` refers to the minimum implementation, wrt both version and flags.

Reply:
so the "min" doesn't really count for flags, right? IMHO the "min" here can be confusing for the reader, when this really is a set of requirements, not all of which has to be "bigger", if you get what I mean.

Reply:
I agree, we should find a different name.
`requirements`?
`request`?
`ext_api` and then document how it should be interpreted?

Reply:
Done (`ext_api`)

Reply:
Oops. Not done, need to fix some more.

Reply:
Done (`request`)